src/Lecture4.hs

```diff
@@ -184,6 +232,17 @@ instance for the 'Stats' type itself.
 -}
 
 instance Semigroup Stats where
+    (<>) :: Stats -> Stats -> Stats
+    x <> y = Stats
+        (statsTotalPositions x <> statsTotalPositions y)
+        (statsTotalSum x       <> statsTotalSum y)
+        (statsAbsoluteMax x    <> statsAbsoluteMax y)
+        (statsAbsoluteMin x    <> statsAbsoluteMin y)
+        (statsSellMax x        <> statsSellMax y)
+        (statsSellMin x        <> statsSellMin y)
+        (statsBuyMax x         <> statsBuyMax y)
+        (statsBuyMin x         <> statsBuyMin y)
```

Comment:
You need to append `Maybe` values strictly to achieve the final streaming approach. If you have `Just x` and `Just y` then the result of their addition with `<>` will still be lazy because `Maybe` has lazy contents.

Here you can implement a custom specialised function for adding only `Maybe` values strictly and use it for all fields with the type `Maybe something`:

```haskell
strictAppend :: Semigroup a => Maybe a -> Maybe a -> Maybe a
strictAppend x y = case x <> y of
    Nothing -> Nothing
    Just !a -> Just a  -- the !a pattern is the important part here
```

Reply:
And sure enough...
```
➜  ~/workinprogress/haskell/beginners/exercises git:(lecture-4-solutions) ✗ time cabal run lecture4 --allow-newer=base --ghc-options=-dynamic -- test/gen/big.csv                                              
Up to date
Total positions        : 150000000
Total final balance    : -2737500000
Biggest absolute cost  : 150
Smallest absolute cost : 1
Max earning            : 100
Min earning            : 1
Max spending           : 150
Min spending           : 7
Longest product name   : Strawberry

cabal run lecture4 --allow-newer=base --ghc-options=-dynamic --   1308.22s user 309.40s system 169% cpu 15:52.15 total
```